homeassistant/components/weheat/config_flow.py

```diff
@@ -0,0 +1,119 @@
+):
+    """Config flow to handle Weheat OAuth2 authentication."""
+
+    DOMAIN = DOMAIN
+
+    reauth_entry: ConfigEntry | None = None
+    _auth_data: dict = {}
+    _discovered_pumps: list[HeatPumpDiscovery.HeatPumpInfo] = []
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    async def async_oauth_create_entry(self, data: dict) -> ConfigFlowResult:
+        """Override the create entry method to change to the step to find the heat pumps."""
+        if self.reauth_entry:
+            # on a reauth, preserve the heat pump info
+            config_entry = self.hass.config_entries.async_get_entry(
+                self.reauth_entry.entry_id
+            )
+            preserved_data: dict = {}
+            if config_entry is not None:
+                preserved_data = config_entry.data.get(HEAT_PUMP_INFO) or {}
+
+            self.hass.config_entries.async_update_entry(
+                self.reauth_entry, data=(data | {HEAT_PUMP_INFO: preserved_data})
+            )
+            await self.hass.config_entries.async_reload(self.reauth_entry.entry_id)
+            return self.async_abort(reason="reauth_successful")
+
+        self._auth_data = data
+
+        return await self.async_step_find_devices()
+
+    async def async_step_find_devices(self, info=None) -> ConfigFlowResult:
+        """Select the heat pump to control.
+
+        Will skip selection if only one heat pump is found.
+        """
+        if info is None or "uuid" not in info:
+            # nothing select, construct list of devices
+            self._discovered_pumps = await self.hass.async_add_executor_job(
+                HeatPumpDiscovery.discover,
+                API_URL,
+                self._auth_data["token"]["access_token"],
+            )
+
+            if len(self._discovered_pumps) == 0:
+                # when there are no pumps, the user is lacking access
+                return self.async_abort(reason="no_devices_found")
+            if len(self._discovered_pumps) == 1:
+                # just select this pump since it is the only one
+                info = {
+                    "uuid": self._discovered_pumps[0].uuid,
+                    HEAT_PUMP_INFO: self._discovered_pumps[0],
+                }
+
+                await self.async_set_unique_id(info["uuid"])
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(
+                    title="Weheat heatpump", data=dict(self._auth_data | info)
+                )
+
+            # show list of pumps
+            heat_pump_dict = {
+                hp.uuid: f"{hp.name} ({hp.sn})" for hp in self._discovered_pumps
+            }
+            return self.async_show_form(
+                step_id="find_devices",
+                data_schema=vol.Schema({vol.Required("uuid"): vol.In(heat_pump_dict)}),
+            )
+
```

Comment:
Sounds better indeed. I will make a sanity check in case the authorized user has access to many heat pumps. This could happen when someone with an installer level account logs in, it would have privileges to see many. Adding 50 heat pumps is likely not what the user wanted to do at that point.

Reply:
Could you point me to an implementation that does this? I'm having a tough time figuring out how to get the config flow to add more than one entry.

Reply:
We don't want to create multiple entries, we want to create 1 entry that represents the account and can add multiple heatpumps. I am also genuinly wondering why installers use the same account

Reply:
Clear. To clarify: Customer have an account with access to typically 1 or 2 heat pumps. Installers have account which have access to all their installed  heat pumps. In an ideal situation the installer would create a customer account for themselves, but if they don't this could happen.